lib/spack/spack/concretize.py

```diff
@@ -362,7 +362,16 @@ def concretize_compiler(self, spec):
         # compiler_for_spec Should think whether this can be more
         # efficient
         def _proper_compiler_style(cspec, aspec):
-            return spack.compilers.compilers_for_spec(cspec, arch_spec=aspec)
+            compilers = spack.compilers.compilers_for_spec(
+                cspec, arch_spec=aspec
+            )
+            # If the spec passed as argument is concrete we want to check
+            # the versions match exactly
+            if (cspec.concrete and compilers and
+                cspec.version != compilers[0].version):
```

Comment:
@alalazo if there are compilers with mixed `x.y` and `x.y.z` versions, should this filter the list instead of just checking one of the elements?

Reply:
I think so. I didn't take into consideration the case where e.g. both `gcc@4.8.5` and `gcc@4.8` are registered compilers.